File: Support/Fuego/Pythia/pythia-0.4/packages/journal/lib/Diagnostic.cc
// -*- C++ -*-
//
//--------------------------------------------------------------------------------
//
//                              Michael A.G. Aivazis
//                       California Institute of Technology
//                       (C) 1998-2003  All Rights Reserved
//
// <LicenseText>
//
//--------------------------------------------------------------------------------
//

#include <portinfo>
#include <map>
#include <vector>
#include <string>
#include <sstream>

#include "Category.h"
#include "Diagnostic.h"
#include "Entry.h"
#include "Device.h"
#include "Journal.h"

using namespace journal;

// interface
Diagnostic::journal_t & Diagnostic::journal() {
    // statics
    static Diagnostic::journal_t * _journal = new Journal();

    return *_journal;
}

void Diagnostic::newline() {

    if (!_category.state()) {
        return;
    }

    _newline();

    return;
}

void Diagnostic::record() {
    if (!_category.state()) {
        return;
    }

    // save the current buffer contents
    _newline();

    // record the journal entry
    journal().record(*_entry);

    // reset the entry
    delete _entry;
    _entry = new Entry;
    (*_entry)["category"] = _name;
    (*_entry)["facility"] = _facility;

    return;
}

void Diagnostic::attribute(Diagnostic::string_t key, Diagnostic::string_t value) {

    (*_entry)[key] = value;
    return;
}


Diagnostic::string_t Diagnostic::str() const {
    return _buffer.str();
}

// meta-methods
Diagnostic::~Diagnostic() {}

Diagnostic::Diagnostic(string_t facility, string_t name, category_t & category):
    _name(name),
    _facility(facility),
    _category(category),
    _buffer(),
    _entry(new Entry)
{
    (*_entry)["category"] = _name;
    (*_entry)["facility"] = _facility;
}

// implementation
void Diagnostic::_newline() {
    _entry->newline(_buffer.str());
    _buffer.str(string_t());

    return;
}

// version
// $Id$

// End of file

File: Support/Fuego/Pythia/pythia-0.5/packages/journal/journal/Diagnostic.py
#!/usr/bin/env python
#
#  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#
#                               Michael A.G. Aivazis
#                        California Institute of Technology
#                        (C) 1998-2003 All Rights Reserved
#
#  <LicenseText>
#
#  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#


from __future__ import absolute_import

import traceback
from builtins import object

import journal

from .Entry import Entry


class Diagnostic(object):
    def line(self, message):
        if not self._state:
            return

        self._entry.line(message)
        return self

    def log(self, message=None):
        if not self._state:
            return

        if message is not None:
            self._entry.line(message)

        stackDepth = -2
        stackTrace = traceback.extract_stack()
        file, line, function, src = stackTrace[stackDepth]

        meta = self._entry.meta
        meta["category"] = self.name
        meta["facility"] = self.facility
        meta["filename"] = file
        meta["function"] = function
        meta["line"] = line
        meta["src"] = src
        meta["stack-trace"] = stackTrace[: stackDepth + 1]

        journal.journal().record(self._entry)

        if self._fatal:
            raise self.Fatal

        self._entry = Entry()
        return self

    def activate(self):
        self._state = True
        return self

    def deactivate(self):
        self._state = False
        return self

    def flip(self):
        self._state ^= True
        return self

    def __init__(self, name, facility, defaultState, fatal=False):
        self.name = name
        self.facility = facility

        self._entry = Entry()
        self._state = defaultState
        self._fatal = fatal

        return

    def _getState(self):
        return self._state

    def _setState(self, state):
        self._state = state
        return

    state = property(_getState, _setState, None, "")

    class Fatal(Exception):
        def __str__(self):
            return "fatal diagnostic"


# version
__id__ = "$Id$"

#  End of file
